### PR TITLE
Dumb checkpointing

### PR DESCRIPTION
**Pull request by wence-**
It looks like I am not going to be able to get "proper" checkpointing done by the end of the week.  So here's the dumb version.  Make a checkpoint file with:

```python
chk = DumbCheckpoint("filename.h5", mode=FILE_CREATE)

```

Then store some function in it

```python
chk.store(function)
```

`function.name()` is used as the path to the data in the hdf5 file.

To later load data do:

```python

chk.load(function2)
```

The data is looked up by name.

For both `load` and `store` you can optionally specify a name, overriding what the function provides.

```python
chk.store(function, name="foo")
chk.load(function2, name="foo")
```

Three modes are available, `FILE_READ`, `FILE_CREATE`, and `FILE_UPDATE`.  The first is read-only, the second creates (if not existing) and truncates otherwise, the last is for read-write access.

Because I only store data (not topology).  It is the user's responsibility to ensure that the mesh/functionspace are the same.  In particular, I raise an error if you try and read in data on a different number of processes from what you dumped on, but the following is also not likely to succeed:

```python

m1 = Mesh(..., reorder=False)

...

f = Function(FS(m1, ...))

chk.store(f)

...

m2 = Mesh(..., reorder=True)

f2 = Function(FS(m2, ...))
chk.load(f2)
```

IOW, you can use this to restart on the same number of processes if you have the same mesh lying around, but otherwise not.

### CODE REVIEW

**Comment by wence-** [comment]
OK, I've cleaned things up a bit, added a checkpoint->h5py File conversion (if PETSc and h5py are linked against the same HDF5 library) ability and added documentation in the manual.  I claim this is good to go.

File: tests/checkpoints/test_dumb_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    f = Function(V, name="f")
+
+    expr = Expression("x[0]*x[1]")
+
+    f.interpolate(expr)
+
+    f2 = Function(V, name="f")
+
+    dumpfile = op2.MPI.comm.bcast(dumpfile, root=0)
+    chk = DumbCheckpoint(dumpfile, mode=FILE_CREATE)
```
Comment (dham) [inline tests/checkpoints/test_dumb_checkpoint.py]: Shouldn't this test (and the others) remove the checkpoint file when it's done?

Comment (wence-) [inline tests/checkpoints/test_dumb_checkpoint.py]: The dumpfile is created in a temporary directory using the pytest tmpdir fixture which is guaranteed to be unique, and is cleaned up in due course by pytest.

**Comment by wence-** [comment]
A comment.  This puts a requirement that the PETSc we use should be built with HDF5 support.  Do I add this to the firedrake-install minimal petsc options.  Also, I forget where else we document dependencies.  I can probably fix things up so that checkpointing is just disabled completely if PETSc doesn't support HDF5 viewers, but that's likely not a good option.

**Comment by wence-** [comment]
Forgot to wire up the tests on travis, so that needs fixing

**Comment by miklos1** [comment]
Fine.

**Comment by tkarna** [comment]
Thanks for implementing this. I think it would make sense to add HDF5 support to firedrake-install --minimal_petsc options so that checkpointing always works.

**Comment by wence-** [comment]
Yes, I am inclined to agree.

**Comment by wence-** [comment]
I have done further implementation + documentation.  I now use h5py for attribute writing and the install script ensures that h5py is linked to the same hdf5 as petsc.  I hope this all works, I have tested on mac with a virtualenv, but not yet linux (hopefully travis will do so).

**Pull request merged**